scripts/busway-config-migrate.jq

```diff
@@ -0,0 +1,88 @@
+.screens += (
+  .screens | with_entries(
+    .value.app_params as $params
+    | .value.app_params.sections as $sections
+    | select(.value.app_id == "solari")
+    | select($params.overhead == false)
+    | {
+      key: "\(.key)-V2",
+      value: (.value + {
+        app_id: "busway_v2",
+        app_params: {
+          header: { stop_name: $params.station_name },
+          evergreen_content: [],
+          departures: {
+            sections: [
+              $sections[] | {
+                header: {
+                  title: (if .name == "" then null else .name end),
+                  arrow,
+                  read_as: (
+                    if .audio.wayfinding == .name then
+                      null
+                    else
+                      .audio.wayfinding
+                    end
+                  )
+                },
+                query,
+                filters: {
+                  max_minutes: .layout | (
+                    if .type == "bidirectional" or
+                        .opts.max_minutes == "infinity" then
+                      null
+                    else
+                      .opts.max_minutes
+                    end
+                  ),
+                  # NOTE: In the old filter format, `direction_id` must be only
+                  # one of the two directions; in the new format, it is allowed
+                  # to be `null`, indicating both directions. We should manually
+                  # combine instances of "route X dir 0" + "route X dir 1".
+                  route_directions: .layout | (
+                    if .type == "upcoming" and
+                        .opts.routes.route_list != [] then
+                      .opts.routes | {
+                        action,
+                        targets: .route_list[]
+                          | { route_id: .route, direction_id }
+                      }
+                    else
+                      null
+                    end
+                  )
+                },
+                layout: (
+                  if .layout.type == "upcoming" then
+                    # NOTE: The v1 config cannot specify a base and max size
+                    # separately; `visible_rows` (if Later Departures enabled)
+                    # or `num_rows` (if not) fills both roles. This sometimes
+                    # means extra space on the screen goes unused. We should
+                    # manually change `max` to `null` on sections where it
+                    # makes sense for them to use any extra space available.
+                    .layout.opts
+                    | (
+                        # `visible_rows` counts Later Departures as one "row".
+                        if .paged then .visible_rows - 1 else .num_rows end
+                      ) as $max
+                    | { min: 1, max: $max, base: $max, include_later: .paged }
+                  else
+                    { min: 1, max: null, base: null, include_later: false }
```

Comment:
Is the `layout` object even needed for bidirectional if all of these values are fixed in that case? Would it make sense for it to just be `layout: null`? Or, `layout: "bidirectional"` and omit the separate `bidirectional` boolean field?

If someone manually changed these values for an existing bidirectional Busway section, e.g. setting `min: 2`, would the logic be able to handle it gracefully?

(Having a value be either an object or a string might be a bit weird, I guess...)

Reply:
In terms of the implementation `bidirectional` is a server-side filter on the departures returned to the client, whereas `layout` controls the client-side behavior when resizing sections to fit. I didn't envision the client looking at `bidirectional` at all. Which brings up a good point: we should migrate existing `bidirectional` sections to have `"min": 2` in the new format, or else it's technically possible for them to get shrunk down to the default minimum of 1 departure, which we probably don't want.

So why is `bidirectional` not part of `Filters`, you ask? Probably just because it was already there in the root `Section` configuration and I didn't think to move it. From what Dave was saying in the Slack threads where we discussed this, we want to develop a more robust and universally-applicable notion of "bidirectional" in the future (which would probably be called something else), so this field may be destined for eventual relocation, renaming, or outright removal — which puts it in a weird spot right now.